storage-node/src/cache/mod.rs

```diff
@@ -9,14 +23,45 @@ pub type ParpulseCacheKey = String;
 /// This is just a prototype and we might refine it later.
 pub type ParpulseCacheValue = (String, usize);
 
-/// [`ParpulseCache`] offers a unified interface for different cache algorithms.
-///
-/// Feel free to modify the methods.
+impl CacheValue for ParpulseCacheValue {
+    type Value = String;
+
+    fn into_value(self) -> Self::Value {
+        self.0
+    }
+
+    fn as_value(&self) -> &Self::Value {
+        &self.0
+    }
+
+    fn size(&self) -> usize {
+        self.1
+    }
+}
+
 pub trait ParpulseCache {
-    // TODO: Consider making the cache lock-free so that we can have `&self`
-    // instead of `&mut self` here, which enables multiple calls to the cache at
-    // the same time.
-    fn get_value(&mut self, key: &ParpulseCacheKey) -> Option<ParpulseCacheValue>;
+    /// Gets a value from the cache. Might has side effect on the cache (e.g.
+    /// modifying some bookkeeping fields in the cache).
+    fn get(&mut self, key: &ParpulseCacheKey) -> Option<&ParpulseCacheValue>;
+
+    /// Puts a value into the cache.
+    fn put(&mut self, key: ParpulseCacheKey, value: ParpulseCacheValue) -> bool;
+
+    /// Returns a reference to the value in the cache with no side effect on the
+    /// cache.
+    fn peek(&self, key: &ParpulseCacheKey) -> Option<&ParpulseCacheValue>;
+
+    /// Returns the number of the objects in the cache.
+    fn len(&self) -> usize;
+
+    /// Returns the total size of the objects in the cache.
+    fn size(&self) -> usize;
+
+    fn is_empty(&self) -> bool;
+
+    fn max_capacity(&self) -> usize;
+
+    fn set_max_capacity(&mut self, capacity: usize);
```

Comment:
Getter & Setter naming convention in Rust: https://github.com/rust-lang/rfcs/blob/master/text/0344-conventions-galore.md#gettersetter-apis